lib/api/src/grpc/transport_channel_pool.rs

```diff
@@ -0,0 +1,68 @@
+use rand::seq::SliceRandom;
+use std::collections::HashMap;
+use std::time::Duration;
+use std::vec::Vec;
+use tonic::transport::{Channel, Error, Uri};
+
+/// Holds a pool of channels established for a set of URIs.
+/// Channel are shared by cloning them.
+/// Make the `pool_size` larger to increase throughput.
+#[derive(Default)]
+pub struct TransportChannelPool {
+    ip_to_pool: tokio::sync::RwLock<HashMap<Uri, Vec<Channel>>>,
+    pool_size: usize,
+    grpc_timeout: Duration,
+}
+
+impl TransportChannelPool {
+    pub fn new(p2p_grpc_timeout: Duration, pool_size: usize) -> Self {
+        Self {
+            ip_to_pool: Default::default(),
+            grpc_timeout: p2p_grpc_timeout,
+            pool_size,
+        }
+    }
+
+    pub async fn make_channel(grpc_timeout: Duration, uri: Uri) -> Result<Channel, Error> {
+        let endpoint = Channel::builder(uri)
+            .timeout(grpc_timeout)
+            .connect_timeout(grpc_timeout)
+            .keep_alive_while_idle(true);
+        // `connect` is using the `Reconnect` network service internally to handle dropped connections
+        endpoint.connect().await
```

Comment:
it is interesting to see how the `ChannelBuilder` is turned into a connection eventually.
One can see the stack of Tower's layers https://github.com/hyperium/tonic/blob/de2e4ac077c076736dc451f3415ea7da1a61a560/tonic/src/transport/service/connection.rs#L30-L73